### PR TITLE
IF: Unification: add latest commit QC known to proposer to block extension (final part)

### PR DESCRIPTION
**Pull request by linh2931**
Continuation of https://github.com/AntelopeIO/leap/pull/2011. Add QC to block extension if needed.

Resolved https://github.com/AntelopeIO/leap/issues/1586

### CODE REVIEW

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -548,8 +548,13 @@ struct assembled_block {
...
                                    },
                                    [&](assembled_block_if& ab) {
+                                      std::optional<quorum_certificate> qc;
+                                      if( ab.qc && ab.bhs.is_needed(*ab.qc ) ) {
+                                         qc = *ab.qc;
+                                      }
+
                                       auto bsp = std::make_shared<block_state>(ab.bhs, std::move(ab.trx_metas),
-                                                                               std::move(ab.trx_receipts));
+                                                                               std::move(ab.trx_receipts), qc);
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/controller.cpp:557]: Actually, I see no reason to copy it at all. Pass the `ab.qc` to `block_state` constructor as `const std::optional<quorum_certificate>&`  and do the `is_needed` check in block_state constructor and pack `*qc` there.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/controller.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp]: Great!

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -549,7 +549,7 @@ struct assembled_block {
                                    },
                                    [&](assembled_block_if& ab) {
                                       auto bsp = std::make_shared<block_state>(ab.bhs, std::move(ab.trx_metas),
-                                                                               std::move(ab.trx_receipts));
+                                                                               std::move(ab.trx_receipts), std::move(ab.qc));
```
Comment (heifner) [inline libraries/chain/controller.cpp:552]: This should not be a `std::move`

Comment (greg7mdp) [reply libraries/chain/controller.cpp]: > This should not be a `std::move`

Why not? I think the constructor should take a rvalue and it should be a std::move.

Comment (heifner) [reply libraries/chain/controller.cpp]: If we cache it in `block_state` then makes sense to move it, otherwise faster to just pass by value.

Comment (linh2931) [reply libraries/chain/controller.cpp]: We don't store it in `bloclk_state`. Do you want me to change to pass by value?

Comment (greg7mdp) [reply libraries/chain/controller.cpp]: You can leave it as is, thanks Lin. We can change it if the QC needs to be stored in block_state.

**Review by greg7mdp — APPROVED** [review]

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Continuing PR #2011 add quorum certificate to block header extension.
Note:end

**Pull request merged**